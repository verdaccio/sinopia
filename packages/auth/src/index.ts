export { Auth } from './auth';
export * from './utils';
export * from './types';
