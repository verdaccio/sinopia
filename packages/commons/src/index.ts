export * from './constants';
export * from './helpers/pkg';
