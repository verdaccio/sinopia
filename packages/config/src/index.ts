export * from './config';
export * from './config-path';
export * from './token';
