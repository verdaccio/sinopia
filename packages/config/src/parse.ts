import fs from 'fs';
import YAML from 'js-yaml';
import { APP_ERROR } from '@verdaccio/commons-api';
import { ConfigRuntime, ConfigYaml } from '@verdaccio/types';

export function parseConfigFile(configPath: string): ConfigRuntime {
  try {
    if (/\.ya?ml$/i.test(configPath)) {
      const yamlConfig = YAML.safeLoad(fs.readFileSync(configPath, 'utf8')) as ConfigYaml;
      return Object.assign({}, yamlConfig, {
        config_path: configPath,
      });
    }

    const jsonConfig = require(configPath) as ConfigYaml;
    return Object.assign({}, jsonConfig, {
      config_path: configPath,
    });
  } catch (e) {
    if (e.code !== 'MODULE_NOT_FOUND') {
      e.message = APP_ERROR.CONFIG_NOT_VALID;
    }

    throw new Error(e);
  }
}
