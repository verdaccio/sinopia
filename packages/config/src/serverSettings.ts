export default {
  // https://www.npmjs.com/package/express-rate-limit
  rateLimit: {
    windowMs: 15 * 60 * 1000,
    max: 10000,
  },
};
