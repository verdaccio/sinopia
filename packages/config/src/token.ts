import { randomBytes } from 'crypto';

// TODO: code duplicated at @verdaccio/signature
export const TOKEN_VALID_LENGTH = 32;

/**
 * Secret key must have 32 characters.
 * // TODO: code duplicated at @verdaccio/signature
 */
export function generateRandomSecretKey(): string {
  return randomBytes(TOKEN_VALID_LENGTH).toString('base64').substring(0, TOKEN_VALID_LENGTH);
}
