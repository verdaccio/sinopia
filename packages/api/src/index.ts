import _ from 'lodash';
import express, { Express } from 'express';

import {
  match,
  validateName,
  validatePackage,
  encodeScopePackage,
  antiLoop,
} from '@verdaccio/middleware';
import { IAuth } from '@verdaccio/auth';
import { IStorageHandler } from '@verdaccio/store';
import { Config } from '@verdaccio/types';
import bodyParser from 'body-parser';

import whoami from './whoami';
import ping from './ping';
import user from './user';
import distTags from './dist-tags';
import publish from './publish';
import search from './search';
import pkg from './package';
import stars from './stars';
import profile from './v1/profile';
import token from './v1/token';
import v1Search from './v1/search';

export default function (
  config: Config,
  auth: IAuth,
  storage: IStorageHandler
): Express.Application {
  /* eslint new-cap:off */
  const app = express.Router();
  /* eslint new-cap:off */

  // validate all of these params as a package name
  // this might be too harsh, so ask if it causes trouble
  app.param('package', validatePackage);
  app.param('filename', validateName);
  app.param('tag', validateName);
  app.param('version', validateName);
  app.param('revision', validateName);
  app.param('token', validateName);

  // these can't be safely put into express url for some reason
  // TODO: For some reason? what reason?
  app.param('_rev', match(/^-rev$/));
  app.param('org_couchdb_user', match(/^org\.couchdb\.user:/));
  app.param('anything', match(/.*/));
  app.use(auth.apiJWTmiddleware());
  app.use(bodyParser.json({ strict: false, limit: config.max_body_size || '10mb' }));
  // @ts-ignore
  app.use(antiLoop(config));
  // encode / in a scoped package name to be matched as a single parameter in routes
  app.use(encodeScopePackage);
  // for "npm whoami"
  whoami(app);
  pkg(app, auth, storage, config);
  profile(app, auth);
  search(app, auth, storage);
  user(app, auth, config);
  distTags(app, auth, storage);
  publish(app, auth, storage, config);
  ping(app);
  stars(app, storage);

  if (config?.flags?.search === true) {
    v1Search(app, auth, storage);
  }

  if (config?.flags?.token === true) {
    token(app, auth, storage, config);
  }
  return app;
}
