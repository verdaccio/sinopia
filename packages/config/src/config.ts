import assert from 'assert';
import _ from 'lodash';
import buildDebug from 'debug';

import {
  getMatchedPackagesSpec,
  normalisePackageAccess,
  sanityCheckUplinksProps,
  uplinkSanityCheck,
  generateRandomHexString,
  getUserAgent,
  isObject,
} from '@verdaccio/utils';
import { APP_ERROR } from '@verdaccio/dev-commons';

import {
  PackageList,
  Config as AppConfig,
  Security,
  Logger,
  PackageAccess,
} from '@verdaccio/types';
import { generateRandomSecretKey } from './token';

const LoggerApi = require('@verdaccio/logger');

const strategicConfigProps = ['uplinks', 'packages'];
const allowedEnvConfig = ['http_proxy', 'https_proxy', 'no_proxy'];

export type MatchedPackage = PackageAccess | void;

export interface StartUpConfig {
  storage: string;
  plugins?: string;
  self_path: string;
}

const debug = buildDebug('verdaccio:config');

/**
 * Coordinates the application configuration
 */
class Config implements AppConfig {
  public logger: Logger;
  public user_agent: string;
  // @ts-ignore
  public secret: string;
  public uplinks: any;
  public packages: PackageList;
  public users: any;
  public server_id: string;
  public self_path: string;
  public storage: string | void;
  public plugins: string | void;
  // @ts-ignore
  public security: Security;

  public constructor(config: StartUpConfig) {
    const self = this;
    this.logger = LoggerApi.logger;
    this.self_path = config.self_path;
    this.storage = config.storage;
    this.plugins = config.plugins;

    for (const configProp in config) {
      if (self[configProp] == null) {
        self[configProp] = config[configProp];
      }
    }

    // @ts-ignore
    if (_.isNil(this.user_agent)) {
      this.user_agent = getUserAgent();
    }

    // some weird shell scripts are valid yaml files parsed as string
    assert(_.isObject(config), APP_ERROR.CONFIG_NOT_VALID);

    // sanity check for strategic config properties
    strategicConfigProps.forEach(function (x): void {
      if (self[x] == null) {
        self[x] = {};
      }

      assert(isObject(self[x]), `CONFIG: bad "${x}" value (object expected)`);
    });

    this.uplinks = sanityCheckUplinksProps(uplinkSanityCheck(this.uplinks));

    if (_.isNil(this.users) === false) {
      this.logger.warn(`[users]: property on configuration file
      is not longer supported, property being ignored`);
    }

    this.packages = normalisePackageAccess(self.packages);

    // loading these from ENV if aren't in config
    allowedEnvConfig.forEach((envConf): void => {
      if (!(envConf in self)) {
        self[envConf] = process.env[envConf] || process.env[envConf.toUpperCase()];
      }
    });

    // unique identifier of self server (or a cluster), used to avoid loops
    // @ts-ignore
    if (!this.server_id) {
      this.server_id = generateRandomHexString(6);
    }
  }

  /**
   * Check for package spec
   */
  public getMatchedPackagesSpec(pkgName: string): MatchedPackage {
    return getMatchedPackagesSpec(pkgName, this.packages);
  }

  /**
   * Store or create whether receive a secret key
   */
  public checkSecretKey(secret: string): string {
    debug('check secret key');
    if (_.isString(secret) && _.isEmpty(secret) === false) {
      this.secret = secret;
      debug('reusing previous key');
      return secret;
    }
    // it generates a secret key
    // FUTURE: this might be an external secret key, perhaps within config file?
    debug('generate a new key');
    this.secret = generateRandomSecretKey();
    return this.secret;
  }
}

export { Config };
