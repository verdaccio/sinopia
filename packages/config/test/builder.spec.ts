import { ConfigBuilder } from '../src';

describe('Config builder', () => {
  test('should create a configuration file as object', () => {
    const config = ConfigBuilder.build();
    config
      .addUplink('upstream', { url: 'https://registry.verdaccio.org' })
      .addUplink('upstream2', { url: 'https://registry.verdaccio.org' })
      .addPackageAccess('upstream/*', {
        access: 'public',
        publish: 'foo, bar',
        unpublish: 'foo, bar',
        proxy: 'some',
      })
      .addLogger({ level: 'info', type: 'stdout', format: 'json' })
      .addAuth({ htpasswd: { file: '.htpasswd' } })
      .addStorage('/tmp/verdaccio')
      .addSecurity({ api: { legacy: true } });
    expect(config.getConfig()).toEqual({
      security: {
        api: {
          legacy: true,
        },
      },
      auth: {
        htpasswd: {
          file: '.htpasswd',
        },
      },
      storage: '/tmp/verdaccio',
      packages: {
        'upstream/*': {
          access: 'public',
          publish: 'foo, bar',
          unpublish: 'foo, bar',
          proxy: 'some',
        },
      },
      uplinks: {
        upstream: {
          url: 'https://registry.verdaccio.org',
        },
        upstream2: {
          url: 'https://registry.verdaccio.org',
        },
      },
      log: {
        level: 'info',
        type: 'stdout',
        format: 'json',
      },
    });
  });

  test('should create a configuration file as yaml', () => {
    const config = ConfigBuilder.build();
    config
      .addUplink('upstream', { url: 'https://registry.verdaccio.org' })
      .addUplink('upstream2', { url: 'https://registry.verdaccio.org' })
      .addPackageAccess('upstream/*', {
        access: 'public',
        publish: 'foo, bar',
        unpublish: 'foo, bar',
        proxy: 'some',
      })
      .addLogger({ level: 'info', type: 'stdout', format: 'json' })
      .addStorage('/tmp/verdaccio')
      .addSecurity({ api: { legacy: true } });
    expect(config.getAsYaml()).toMatchSnapshot();

    expect(config.getConfig()).toEqual({
      uplinks: {
        upstream: {
          url: 'https://registry.verdaccio.org',
        },
        upstream2: {
          url: 'https://registry.verdaccio.org',
        },
      },
      packages: {
        'upstream/*': {
          access: 'public',
          publish: 'foo, bar',
          unpublish: 'foo, bar',
          proxy: 'some',
        },
      },
      security: {
        api: {
          legacy: true,
        },
      },
      log: {
        level: 'info',
        type: 'stdout',
        format: 'json',
      },
      storage: '/tmp/verdaccio',
    });
  });

  test('should merge configurations', () => {
    // @ts-expect-error
    const config = ConfigBuilder.build({ security: { api: { legacy: false } } });
    config.addSecurity({ web: { verify: {}, sign: { algorithm: 'ES256' } } });
    config.addStorage('/tmp/verdaccio');
    expect(config.getConfig()).toEqual({
      security: {
        api: {
          legacy: false,
        },
        web: {
          verify: {},
          sign: {
            algorithm: 'ES256',
          },
        },
      },
      uplinks: {},
      packages: {},
      storage: '/tmp/verdaccio',
    });
  });
});
