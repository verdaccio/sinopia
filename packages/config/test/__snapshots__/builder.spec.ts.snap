// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Config builder > should create a configuration file as yaml 1`] = `
"uplinks:
  upstream:
    url: https://registry.verdaccio.org
  upstream2:
    url: https://registry.verdaccio.org
packages:
  upstream/*:
    access: public
    publish: foo, bar
    unpublish: foo, bar
    proxy: some
security:
  api:
    legacy: true
log:
  level: info
  type: stdout
  format: json
storage: /tmp/verdaccio
"
`;
